File: fix_v324.js
// Final fix: remove duplicate <div class="tl"> and orphaned h+=
const fs = require('fs');
const path = 'C:/Users/Server/.openclaw/workspace/Shop_vnext/src/Index.html';

let content = fs.readFileSync(path, 'utf8');

// Check the grid structure
const lines = content.split('\n');

// Line 212 (index 211): lowStock tile - ends with </div></div>'; (tile close + grid close)
// Line 213 (index 212): h+='<div class="tl">'; - duplicate grid open
// Line 214 (index 213): pendingPhotos tile 
// Line 215 (index 214): processPhotos tile
// Line 216 (index 215): h+='</div>'; - duplicate grid close

// The issue: Line 212's </div></div> closes both the tile AND the entire grid.
// Then Line 213 opens a new grid, Line 214-215 add tiles, Line 216 closes that new grid.

// FIX 1: Delete the orphaned h+= on line 182
// It's between function se() and function sr()
// Find it by searching for the pattern
let orphanIdx = -1;
for (let i = 180; i < 185; i++) {
  if (lines[i] && lines[i].trim().startsWith('h+=') && lines[i].includes('addNote') && !lines[i].trim().startsWith('function')) {
    orphanIdx = i;
    break;
  }
}

if (orphanIdx >= 0) {
  console.log(`[FIX 1] Deleting orphaned h+= on line ${orphanIdx + 1}`);
  console.log(`  Content: ${lines[orphanIdx].trim().substring(0, 80)}`);
  lines[orphanIdx] = ''; // will filter out empty lines
}

// FIX 2: Merge grids - make ONE 6-tile grid instead of two
// Step A: On line 212 (index 211), change </div></div>' to </div>' 
// This removes the grid close, keeps tile close
if (lines[211]) {
  const before = lines[211];
  lines[211] = before.replace('</div></div>', '</div>');
  if (before !== lines[211]) {
    console.log(`[FIX 2a] Line 212: Removed grid-close </div>, keeping tile-close </div>`);
    console.log(`  Before: ${before.trim().substring(0, 80)}`);
    console.log(`  After:  ${lines[211].trim().substring(0, 80)}`);
  }
}

// Step B: Delete line 213 (index 212): h+='<div class="tl">';
if (lines[212] && lines[212].includes("h+='<div class=\"tl\">'")) {
  console.log(`[FIX 2b] Line 213: Deleted duplicate grid open`);
  lines[212] = '';
}

// Step C: Delete line 216 (index 215): h+='</div>'; that closes the duplicate grid
if (lines[215] && lines[215].trim() === "h+='</div>';") {
  console.log(`[FIX 2c] Line 216: Deleted duplicate grid close`);
  lines[215] = '';
}

// FIX 3: Typo รอดำเนินการ → รอดำเนินการ
let typoCount = 0;
for (let i = 0; i < lines.length; i++) {
  if (lines[i].includes("รอดาเนินการ")) {
    lines[i] = lines[i].replace(/รอดาเนินการ/g, 'รอดำเนินการ');
    typoCount++;
    console.log(`[FIX 3] Line ${i+1}: Fixed typo รอดำเนินการ → รอดำเนินการ`);
  }
}
if (typoCount === 0) {
  console.log(`[FIX 3] No typo found (already fixed or different)`);
  // Check what's there
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].includes("รอด") && lines[i].includes("ดำเนินการ") && lines[i].includes("cSt")) {
      console.log(`  Line ${i+1} has cSt with รอ text: ${lines[i].substring(0, 100)}`);
    }
  }
}

// REBUILD
content = lines.filter(l => l !== '').join('\n');

// SAVE
fs.writeFileSync(path, content, 'utf8');
console.log(`\nSaved: ${content.length} bytes`);

// VERIFY
const check = fs.readFileSync(path, 'utf8');
const m = check.match(/<script>([\s\S]*?)<\/script>/);
if (!m) {
  console.log('❌ NO SCRIPT TAG FOUND');
} else {
  try {
    new Function(m[1]);
    console.log('✅ V324 JS SYNTAX VALID');
  } catch(e) {
    console.log('❌ JS ERROR:', e.message);
    const lineMatch = e.message.match(/line (\d+)/);
    if (lineMatch) {
      const jsLine = parseInt(lineMatch[1]);
      const jsLines = m[1].split('\n');
      const start = Math.max(0, jsLine - 3);
      const end = Math.min(jsLines.length, jsLine + 3);
      for (let i = start; i < end; i++) {
        const mark = (i === jsLine - 1) ? '>>>' : '   ';
        console.log(`${mark} JS Line ${i+1}: ${jsLines[i].substring(0, 120)}`);
      }
    }
  }
}

// Confirm grid tile count
const tlMatch = content.match(/function render\([\s\S]*?sr\(h\)/);
if (tlMatch) {
  const tiles = (tlMatch[0].match(/h\+='\<div class="tt"/g) || []).length;
  console.log(`Render function has ${tiles} tiles (expected 6)`);
}

File: _fix_v324.py
# -*- coding: utf-8 -*-
"""V324 Fix - 3 surgical edits to Index.html"""
import io, sys, os, subprocess

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

path = r"C:\Users\Server\.openclaw\workspace\Shop_vnext\src\Index.html"

with open(path, "r", encoding="utf-8") as f:
    content = f.read()

# ========== SHOW CURRENT STATE ==========
print("=== Checking Index.html ===\n")

# Check for the orphaned h+= line between se() and sr()
orphan = 'h+=\'<div class="b2 ba" onclick="addNote'
orphan_pos = content.find(orphan)
print(f"1. Orphaned h+= addNote: found at position {orphan_pos}")

# Check for duplicate grid
dup_tl = content.count('h+=\'<div class="tl">\'')
print(f"2. Duplicate <div class=\"tl\">: {dup_tl} occurrences (should be 1)")

# Check typo
typo_check = content.count("รอดาเนินการ")
print(f"3. Typo รอดาเนินการ (missing า): {typo_check} occurrences")

# ========== FIX ==========

# FIX 1: Remove orphaned h+= line
# This line floats between se() closing and sr() opening — delete it
content = content.replace(
    '  h+=\'<div class="b2 ba" onclick="addNote(\\\'\'\+ id +\'\\\')">💾 บันทึกหมายเหตุ</div>\';\nfunction sr',
    'function sr'
)
# Also without trailing newline variant
content = content.replace(
    '  h+=\'<div class="b2 ba" onclick="addNote(\\\'\'\+ id +\'\\\')">💾 บันทึกหมายเหตุ</div>\';\n',
    ''
)

# FIX 2: Fix duplicate grid in render()
# The render() function creates stats grid with closing </div></div>';
# Then immediately opens another <div class="tl"> for photo tiles 
# → Merge into ONE grid: close lowStock tile + add 2 photo + close grid

old_dup = """h+='<div class="tt"><div class="n c4">'+(s.lowStock||0)+'</div><div class="l">🚨 สต็อกต่ำ</div></div></div>';
    h+='<div class="tl">';
    h+='<div class="tt"><div class="n" style="color:#6C63FF">'+(s.pendingPhotos||0)+'</div><div class="l">📸 รอดำเนินการ</div></div>';
    h+='<div class="tt"><div class="l" style="color:#6C63FF;cursor:pointer" onclick="processPhotos()">▶️ จัดการรูป</div></div>';
    h+='</div>';"""

new_merged = """h+='<div class="tt"><div class="n c4">'+(s.lowStock||0)+'</div><div class="l">🚨 สต็อกต่ำ</div></div>';
    h+='<div class="tt"><div class="n" style="color:#6C63FF">'+(s.pendingPhotos||0)+'</div><div class="l">📸 รอดำเนินการ</div></div>';
    h+='<div class="tt"><div class="l" style="color:#6C63FF;cursor:pointer" onclick="processPhotos()">▶️ จัดการรูป</div></div>';
    h+='</div>';"""

if old_dup in content:
    content = content.replace(old_dup, new_merged)
    print("\n[Fix] Merged duplicate grid into single 6-tile grid")
else:
    print("\n[Warn] Could not find exact duplicate grid pattern")
    # Try to find what's actually there
    idx = content.find("pendingPhotos")
    if idx > 0:
        ctx_start = max(0, idx - 200)
        ctx_end = min(len(content), idx + 300)
        print("Context around 'pendingPhotos':")
        print(repr(content[ctx_start:ctx_end]))

# FIX 3: Typo รอดาเนินการ → รอดำเนินการ
content = content.replace("'รอดาเนินการ'", "'รอดำเนินการ'")
print("[Fix] Typo: รอดำเนินการ → รอดำเนินการ")

# ========== SAVE ==========
with open(path, "w", encoding="utf-8") as f:
    f.write(content)

print(f"\nSaved: {os.path.getsize(path)} bytes")

# ========== VERIFY ==========
r = subprocess.run(
    ["node", "-e",
     "const fs=require('fs');const c=fs.readFileSync('C:/Users/Server/.openclaw/workspace/Shop_vnext/src/Index.html','utf8');"
     "const m=c.match(/<script>([\\s\\S]*?)<\\/script>/);"
     "if(!m){console.log('NO SCRIPT FOUND');process.exit(1)}"
     "try{new Function(m[1]);console.log('JS SYNTAX VALID ✓')}catch(e){console.log('JS ERROR:',e.message);const lm=e.toString().match(/line (\\d+)/);if(lm){const idx=parseInt(lm[1])-1;const jsLines=m[1].split('\\n');for(let i=Math.max(0,idx-2);i<=Math.min(jsLines.length-1,idx+2);i++){console.log((i===idx?'=>> ':'    ')+'Line '+(i+1)+': '+jsLines[i].substring(0,120))}}}"],
    capture_output=True, text=True
)
print("\n" + r.stdout.strip())
if r.stderr:
    print("STDERR: " + r.stderr[:300])
